### PR TITLE
compiler: refine wasm types proposal

### PR DESCRIPTION
**Pull request by ydnar**
This PR adds tests for the relaxed wasm param and result type restrictions, and re-adds `string` as an allowed param type.

See https://github.com/golang/go/issues/66984 for additional context.

### CODE REVIEW

**Comment by ydnar** [comment]
Looks like I need to rebase!

File: compiler/symbol.go [hunk 1]
```diff
@@ -380,7 +380,7 @@ func (c *compilerContext) checkWasmImport(f *ssa.Function, pragma string) {
 //
 // This previously reflected the additional restrictions documented here:
 // https://github.com/golang/go/issues/59149
-func isValidWasmType(typ types.Type, isPointerOrField bool) bool {
+func isValidWasmType(typ types.Type, isResult, isPointerOrField bool) bool {
```
Comment (dgryski) [inline compiler/symbol.go]: Ugh, multiple bools.  I wonder if we could change this so the caller had something like `isValidWasmType(typ, wasmtypeIsResult|wasmtypeIsPointerOrField)` to make it explicit.  (OTOH, only used in this package, so maybe we're safe?)

File: compiler/testdata/errors.go [hunk 2]
```diff
@@ -13,18 +13,62 @@ func implementation() {
...
+// ERROR: //go:wasmimport modulename invalidparam: unsupported parameter type []byte
+// ERROR: //go:wasmimport modulename invalidparam: unsupported parameter type struct{a int}
+// ERROR: //go:wasmimport modulename invalidparam: unsupported parameter type chan struct{}
+// ERROR: //go:wasmimport modulename invalidparam: unsupported parameter type func()
+//
+//go:wasmimport modulename invalidparam
+func invalidparam(a [4]uint32, b []byte, c struct{ a int }, d chan struct{}, e func())
+
+//go:wasmimport modulename validreturn1
+func validreturn1() int32
```
Comment (dgryski) [inline compiler/testdata/errors.go]: Do we care about having these. named `validreturnInt32` or something so we don't just have numbers?

**Review by dgryski — APPROVED** [review]
LGTM.  Comments are minor and and can be cleaned up later if we care.

**Comment by ydnar** [comment]
Cleaned up!

**Comment by dgryski** [comment]
Merging. I'll fix the CI failure which looks like broken build tags on my end.

**Pull request merged**